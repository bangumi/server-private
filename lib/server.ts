import fastifyRequestContextPlugin from '@fastify/request-context';
import type { Static } from '@sinclair/typebox';
import type Ajv from 'ajv';
import addFormats from 'ajv-formats';
import { DrizzleError } from 'drizzle-orm';
import type { FastifyInstance, FastifyRequest, FastifyServerOptions } from 'fastify';
import { fastify } from 'fastify';
import type { FastifySchemaValidationError } from 'fastify/types/schema.d.ts';
import metricsPlugin from 'fastify-metrics';
import mercurius from 'mercurius';
import { TypeORMError } from 'typeorm';

import { fastifyAltairPlugin } from '@app/lib/graphql/ui.ts';
import * as routes from '@app/routes/index.ts';

import { emptyAuth } from './auth/index.ts';
import * as auth from './auth/index.ts';
import config, { testing, VERSION } from './config.ts';
import type { Context } from './graphql/context.ts';
import { resolvers, schema } from './graphql/schema.ts';
import { repo } from './orm/index.ts';
import type * as res from './types/res.ts';

export function defaultSchemaErrorFormatter(
  errors: FastifySchemaValidationError[],
  dataVar: string,
): Error & Static<typeof res.Error> {
  let text = '';
  const separator = ', ';

  for (const e of errors) {
    text += `${dataVar}${e.instancePath} ${e.message ?? ''}${separator}`;
  }
  return new ValidationError(text.slice(0, -separator.length));
}

class ValidationError extends Error {
  code = 'REQUEST_VALIDATION_ERROR';
  error = 'Bad Request';
  statusCode = 400;

  constructor(msg: string) {
    super(msg);
    this.message = msg;
  }
}

declare module '@fastify/request-context' {
  interface RequestContextData {
    req: unknown;
    user?: number;
  }
}

export async function createServer(
  opts: Omit<
    FastifyServerOptions,
    'onProtoPoisoning' | 'ajv' | 'schemaErrorFormatter' | 'requestIdHeader'
  > = {},
): Promise<FastifyInstance> {
  const ajv: FastifyServerOptions['ajv'] = {
    plugins: [
      function (ajv: Ajv) {
        ajv.addKeyword({ keyword: 'x-examples' });
        ajv.addKeyword({ keyword: 'x-ms-enum' });
        ajv.addKeyword({ keyword: 'x-enum-varnames' });
      },
      addFormats,
    ],
  };

  const server = fastify({
    ...opts,
    onProtoPoisoning: 'error',
    ajv,
    bodyLimit: 8 * 1024 * 1024,
    requestIdHeader: config.server.requestIDHeader,
    schemaErrorFormatter: defaultSchemaErrorFormatter,
  });

  server.setErrorHandler(function (error, request, reply) {
    // hide TypeORM message
    if (error instanceof TypeORMError || error instanceof DrizzleError) {
      this.log.error(error);
      void reply.status(500).send({
        error: 'Internal Server Error',
        message: 'internal database error, please contact admin',
        statusCode: 500,
      });
    } else {
      if (typeof error.statusCode !== 'number' || error.statusCode === 500) {
        this.log.error(error);
        void reply.status(500).send({
          error: 'Internal Server Error',
          message: 'internal error, please contact admin',
          statusCode: 500,
        });
      } else {
        void reply.send(error);
      }
    }
  });

  server.addHook('onRequest', (req, res, done) => {
    void res.header('x-server-version', VERSION);
    done();
  });

  const clientIpHeader = config.server.clientIpHeader;

  server.decorateRequest('ip', {
    getter: function (this: FastifyRequest): string {
      const hRealIp = this.headers[clientIpHeader] as string | undefined;
      if (hRealIp) {
        return hRealIp;
      }

      return this.socket.remoteAddress ?? '127.0.0.1';
    },
  });

  if (!testing) {
    await server.register(metricsPlugin, {
      routeMetrics: {
        groupStatusCodes: false,
        overrides: {
          histogram: {
            buckets: [0.05, 0.1, 0.2, 0.3, 0.5, 0.75, 1, 2, 3],
          },
        },
      },
    });
  }

  await server.register(fastifyRequestContextPlugin, {
    defaultStoreValues: (req) => ({
      req: {
        id: req.id,
        url: req.url,
      },
    }),
  });

  await server.register(mercurius, {
    schema,
    resolvers,
    path: '/v0/graphql',
    jit: 200,
    queryDepth: 5,
    allowBatchedQueries: true,
    context: async (request: FastifyRequest): Promise<Context> => {
      const a = await auth.byHeader(request.headers.authorization);
      if (a) {
        return { repo, auth: a };
      }

      return { repo, auth: emptyAuth() };
    },
  });

  await server.register(fastifyAltairPlugin, {
    path: '/v0/altair/',
    baseURL: '/v0/altair/',
    endpointURL: '/v0/graphql',
    initialSettings: {
      theme: 'dark',
      'plugin.list': ['altair-graphql-plugin-graphql-explorer'],
    },
  });

  await server.register(routes.setup);

  return server;
}
