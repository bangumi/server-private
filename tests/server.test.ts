// remember to wrap all graphql string with gql,
// so vscode, webstorm and prettier will know the string content
// is graphql and support language specific features
import { gql } from 'graphql-tag';
import { createMercuriusTestClient } from 'mercurius-integration-testing';
import { describe, expect, test } from 'vitest';

import { createServer } from '@app/lib/server.ts';

const testClient = createMercuriusTestClient(await createServer(), { url: '/v0/graphql' });

describe('graphql', () => {
  test('should return current user', async () => {
    await expect(
      testClient.query(gql`
        query {
          me {
            id
          }
        }
      `),
    ).resolves.toEqual({
      data: {
        me: null,
      },
    });
  });

  test('should return current user', async () => {
    await expect(
      testClient.query(
        gql`
          query {
            me {
              id
              username
              nickname
            }
          }
        `,
        {
          headers: { authorization: 'Bearer a_development_access_token' },
        },
      ),
    ).resolves.toEqual({
      data: {
        me: { id: 382951, username: '382951', nickname: '树洞酱' },
      },
    });
  });

  test('should return error', async () => {
    await expect(
      testClient.query(
        gql`
          query {
            me {
              id
            }
          }
        `,
        {
          headers: { authorization: 'Bearer a' },
        },
      ),
    ).resolves.toMatchSnapshot();
  });
});

describe('subject', () => {
  test('should get', async () => {
    await expect(
      testClient.query(gql`
        query {
          subject(id: 8) {
            name_cn
          }
        }
      `),
    ).resolves.toEqual({
      data: {
        subject: { name_cn: 'Code Geass 反叛的鲁路修R2' },
      },
    });
  });

  test('should not found', async () => {
    await expect(
      testClient.query(gql`
        query {
          subject(id: 4040404) {
            name_cn
          }
        }
      `),
    ).resolves.toEqual({
      data: {
        subject: null,
      },
    });
  });

  test('should return object for authorized request', async () => {
    await expect(
      testClient.query(
        gql`
          query {
            subject(id: 16) {
              id
            }
          }
        `,
        {
          headers: { authorization: 'Bearer a_development_access_token' },
        },
      ),
    ).resolves.toEqual({
      data: {
        subject: { id: 16 },
      },
    });
  });

  test('should return null for non-authorized request', async () => {
    await expect(
      testClient.query(gql`
        query {
          subject(id: 16) {
            name_cn
          }
        }
      `),
    ).resolves.toEqual({
      data: {
        subject: null,
      },
    });
  });

  test('should get episodes, limit', async () => {
    const query = await testClient.query(gql`
      query {
        subject(id: 8) {
          id
          episodes(limit: 3) {
            id
          }
        }
      }
    `);

    expect(query.data.subject.id).toBe(8);
    expect(query.data.subject.episodes).toHaveLength(3);
  });

  test('should get episodes, -offset', async () => {
    const query = await testClient.query(gql`
      query {
        subject(id: 8) {
          id
          episodes(offset: -3) {
            id
          }
        }
      }
    `);

    expect(query.data.subject.id).toBe(8);
    expect(query.data.subject.episodes).toHaveLength(3);
  });

  test('should subject tags', async () => {
    const query = await testClient.query(gql`
      query {
        subject(id: 8) {
          tags(limit: 3) {
            name
            count
          }
        }
      }
    `);

    expect(query.data.subject).toMatchSnapshot();
  });

  test('should subject relations', async () => {
    const query = await testClient.query(gql`
      query {
        subject(id: 12) {
          relations(limit: 1) {
            relation
            subject {
              id
              name
            }
          }
        }
      }
    `);

    expect(query.data.subject.relations[0].relation).toBe(1004);
    expect(query.data.subject.relations[0].subject.id).toBe(11);
  });

  test('should subject relations with exclude', async () => {
    const query = await testClient.query(gql`
      query {
        subject(id: 12) {
          relations(limit: 1, excludeTypes: [1004]) {
            relation
            subject {
              id
              name
            }
          }
        }
      }
    `);

    expect(query.data.subject.relations).toHaveLength(0);
  });

  test('should reject nested subject relations', async () => {
    const query = await testClient.query(gql`
      query {
        subject(id: 12) {
          relations(limit: 1) {
            relation
            subject {
              id
              name
              relations(limit: 1) {
                relation
                subject {
                  id
                  name
                }
              }
            }
          }
        }
      }
    `);

    expect(query).toMatchSnapshot();
  });
});

describe('character', () => {
  test('should get', async () => {
    await expect(
      testClient.query(gql`
        query {
          character(id: 1) {
            name
            role
          }
        }
      `),
    ).resolves.toEqual({
      data: {
        character: { name: 'ルルーシュ・ランペルージ', role: 1 },
      },
    });
  });
});

describe('person', () => {
  test('should get', async () => {
    await expect(
      testClient.query(gql`
        query {
          person(id: 1) {
            name
            career
          }
        }
      `),
    ).resolves.toEqual({
      data: {
        person: {
          name: '水樹奈々',
          career: ['artist', 'seiyu'],
        },
      },
    });
  });
});
