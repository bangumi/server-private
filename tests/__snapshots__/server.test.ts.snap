// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`graphql > should return error 1`] = `
Object {
  "code": "TOKEN_INVALID",
  "error": "Unauthorized",
  "message": "can't find access token or it has been expired",
  "statusCode": 401,
}
`;

exports[`subject > should reject nested subject relations 1`] = `
Object {
  "data": null,
  "errors": Array [
    Object {
      "message": "\`unnamedQuery query depth (6) exceeds the query depth limit of 4\`",
    },
  ],
}
`;

exports[`subject > should subject tags 1`] = `
Object {
  "tags": Array [
    Object {
      "count": 1645,
      "name": "叛逆的鲁鲁修",
    },
    Object {
      "count": 1229,
      "name": "SUNRISE",
    },
    Object {
      "count": 936,
      "name": "反逆のルルーシュ",
    },
  ],
}
`;
